src/orc2/mod.rs

```diff
@@ -0,0 +1,2186 @@
+        let mut ptr = ptr::null_mut();
+        LLVMError::new(unsafe {
+            LLVMOrcMaterializationResponsibilityDelegate(
+                self.materialization_responsibility,
+                symbols.as_mut_ptr() as *mut LLVMOrcSymbolStringPoolEntryRef,
+                symbols.len(),
+                &mut ptr,
+            )
+        })?;
+        Ok(unsafe { MaterializationResponsibility::new(ptr) })
+    }
+
+    /// Adds dependencies to a symbol `name` that the `MaterializationResponsibility`.
+    pub fn add_dependencies(
+        &self,
+        name: SymbolStringPoolEntry,
+        mut dependencies: DependenceMapPairs,
+    ) {
+        unsafe {
+            LLVMOrcMaterializationResponsibilityAddDependencies(
+                self.materialization_responsibility,
+                name.entry,
+                dependencies.raw_ptr(),
+                dependencies.len(),
+            )
+        }
+    }
+
+    /// Adds dependencies to all symbols of the `MaterializationResponsibility`.
+    pub fn add_dependencies_for_all(&self, mut dependencies: DependenceMapPairs) {
+        unsafe {
+            LLVMOrcMaterializationResponsibilityAddDependenciesForAll(
+                self.materialization_responsibility,
+                dependencies.raw_ptr(),
+                dependencies.len(),
+            );
+        }
+    }
+}
+
+#[llvm_versions(11.0)]
+type LLVMOrcDefinitionGeneratorRef = LLVMOrcJITDylibDefinitionGeneratorRef;
+#[llvm_versions(11.0)]
+unsafe extern "C" fn LLVMOrcDisposeDefinitionGenerator(
+    definition_generator: LLVMOrcDefinitionGeneratorRef,
+) {
+    LLVMOrcDisposeJITDylibDefinitionGenerator(definition_generator);
+}
+
+pub struct DefinitionGeneratorRef<'jit> {
+    definition_generator: LLVMOrcDefinitionGeneratorRef,
+    _marker: PhantomData<&'jit ()>,
+}
+
+impl<'jit> DefinitionGeneratorRef<'jit> {
+    unsafe fn new(definition_generator: LLVMOrcDefinitionGeneratorRef) -> Self {
+        assert!(!definition_generator.is_null());
+        DefinitionGeneratorRef {
+            definition_generator,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl fmt::Debug for DefinitionGeneratorRef<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("DefinitionGenerator")
+            .field("definition_generator", &self.definition_generator)
+            .finish()
+    }
+}
+
+#[derive(Debug)]
+pub struct DefinitionGenerator<'jit> {
+    definition_generator: DefinitionGeneratorRef<'jit>,
+}
+
+impl<'jit> DefinitionGenerator<'jit> {
+    unsafe fn new(definition_generator: LLVMOrcDefinitionGeneratorRef) -> Self {
+        DefinitionGenerator {
+            definition_generator: DefinitionGeneratorRef::new(definition_generator),
+        }
+    }
+
+    #[llvm_versions(12.0..=latest)]
+    pub fn create_custom_capi_definition_generator<G>(
+        capi_definition_generator: &'jit mut Wrapper<'jit, G>,
+    ) -> Self
+    where
+        G: CAPIDefinitionGenerator,
```

Comment:
Back to this, I'm still very confused why it allows dropping `capi_definition_generator` before the instance of `LLVMJIT<'jit>` (when using `&'jit mut G` without `Wrapper`).

I can model that problem in a small reproduction and ask what the guys in the Rust forums think